src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_CAPACITY = 10;
+    private int arraySize;
+    private Object[] values;
+
+    public ArrayList() {
+        arraySize = 0;
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        if (isNotEnoughSpace()) {
+            resize();
+        }
+        values[arraySize] = value;
+        arraySize++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > arraySize || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("wrong index");
+        }
+        if (isNotEnoughSpace()) {
+            resize();
+        }
+        System.arraycopy(values, index, values, index + 1, arraySize - index);
+        values[index] = value;
+        arraySize++;
     }
 
     @Override
     public void addAll(List<T> list) {
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!(index >= arraySize || index < 0)) {
+            return (T) values[index];
+        }
+        throw new ArrayIndexOutOfBoundsException("we have not this place in our list");
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!(index < arraySize && index >= 0)) {
+            throw new ArrayIndexOutOfBoundsException("we have not this place in our list");
+        }
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < arraySize && index >= 0) {
+            T removed = (T) values[index];
+            int numMoved = arraySize - index - 1;
+            System.arraycopy(values, index + 1, values, index, numMoved);
+            values[--arraySize] = null;
+            return removed;
+        }
+        throw new ArrayIndexOutOfBoundsException("wrong index");
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < values.length; i++) {
+            if ((T) values[i] == t || t != null && t.equals((T) values[i])) {
```

Comment:
```suggestion
            if (values[i] == t || t != null && t.equals(values[i])) {
```
value is still an instance of type T, it doesn't matter that it is stored in Object[]